File: src/compare.py
import json
import time
import torch
import sys
import numpy as np
from transformers import pipeline, CLIPProcessor, CLIPModel
from PIL import Image

# Load config + data
with open("config.json") as f:
    config = json.load(f)

with open("data/data.json") as f:
    data = json.load(f)

# Load model + processor
model = CLIPModel.from_pretrained("openai/clip-vit-base-patch32")
processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")

# Load image
image = Image.open("./capture--auto.jpg")

with torch.no_grad():
    inputs = processor(images=image, return_tensors="pt") #
    # get_image_features applies the projection to 512-d
    image_features = model.get_image_features(**inputs) 

# 4. Normalize and Convert to Numpy
image_features = image_features / image_features.norm(p=2, dim=-1, keepdim=True) #
photo_embedding = image_features.numpy().flatten() # Shape (1, 512)

def cosine_similarity(a, b):
    a = np.array(a)
    b = np.array(b)
    return np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b))

best_match = None
best_score = -float("inf")

for release in data:
    score = cosine_similarity(photo_embedding, release["embedding"])

    if score > best_score:
        best_score = score
        best_match = f'{release["id"]} = {release["artist"]} – {release["title"]}'

sys.stdout.write(best_match)
sys.stdout.flush()
sys.exit(0)

File: src/compare.js
// This script loops through all covers and extracts data from them
// It saves this in extractions json
import { pipeline, RawImage } from '@huggingface/transformers';
import embedding from "../data/embedding.json" with { type: 'json' };
import config from '../config.json' with { type: 'json' };

const extractor = await pipeline(
  'image-feature-extraction',
  config.model
);

const cosineSimilarity = (a,b) => {
  let dot = 0, normA = 0, normB = 0;

  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] ** 2;
    normB += b[i] ** 2;
  }

  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

(async() => {
  const photo = await RawImage.read(`./capture.jpg`);
  let photoEmbedding = await extractor(photo, {
    pooling: 'mean',
    normalize: true
  });

  photoEmbedding = photoEmbedding.data;

  let bestMatch = null;
  let bestScore = -Infinity;

  for (const embed of embedding) {
    const score = cosineSimilarity(photoEmbedding, embed.embedding);

    if (score > bestScore) {
      bestScore = score;
      bestMatch = embed;
    }
  }

  console.log(bestMatch.artist, bestMatch.title);
})();